Products/CMFPlone/Portal.py

```diff
@@ -139,6 +139,24 @@ def __before_publishing_traverse__(self, arg1, arg2=None):
 
         super(PloneSite, self).__before_publishing_traverse__(arg1, arg2)
 
+    # Partly from OFS.OrderSupport
+    @security.protected(permissions.AccessContentsInformation)
+    def tpValues(self):
+        # Return a list of subobjects, used by ZMI tree tag (and only there).
+        # see also https://github.com/plone/Products.CMFPlone/issues/3323
+        result = []
+        obj_ids = self.objectIds()
+        obj_ids.sort()
+        getOb = self._getOb
+        for id in obj_ids:
+            obj = getOb(id)
+            if (
+                hasattr(aq_base(obj), 'isPrincipiaFolderish')
+                and obj.isPrincipiaFolderish
+            ):
+                result.append(obj)
+        return result
```

Comment:
Untested, but it seems to me that something like this is preferable:
```
return sorted(
    (obj for obj in self.objectValues() if getattr(aq_base(obj), 'isPrincipiaFolderish', False),
    key=lambda obj: obj.getId(),
)
```
1. you do not append to a list
2. you save an hasattr
3. you do not fetch the objects from the DB one at a time

What do you think?

Reply:
That looks good. You may want to suggest a PR for OFS.ObjetcManager as well!

OT: you now there's a nice Github suggestion feature where I just would need to click on accept ;) ? Lazy me.